paasta_tools/autoscaling/autoscaling_cluster_lib.py

```diff
@@ -49,518 +46,589 @@
+        current_capacity = len(self.asg['Instances'])
+        if current_capacity == 0:
+            new_capacity = int(min(
+                max(1, self.resource['min_capacity']),
+                self.resource['max_capacity']))
+            ideal_capacity = new_capacity
+        else:
+            ideal_capacity = int(ceil((1 + error) * current_capacity))
+            new_capacity = int(min(
+                max(
+                    self.resource['min_capacity'],
+                    floor(current_capacity * (1.00 - MAX_CLUSTER_DELTA)),
+                    ideal_capacity,
+                    0,
+                ),
+                ceil(current_capacity * (1.00 + MAX_CLUSTER_DELTA)),
+                # if max and min set to 0 we still drain gradually
+                max(self.resource['max_capacity'], floor(current_capacity * (1.00 - MAX_CLUSTER_DELTA)))
+            ))
+        log.debug("The ideal capacity to scale to is %d instances" % ideal_capacity)
+        log.debug("The capacity we will scale to is %d instances" % new_capacity)
+        if ideal_capacity > self.resource['max_capacity']:
+            log.warning("Our ideal capacity (%d) is higher than max_capacity (%d). Consider rasing max_capacity!" % (
+                ideal_capacity, self.resource['max_capacity']))
+        if ideal_capacity < self.resource['min_capacity']:
+            log.warning("Our ideal capacity (%d) is lower than min_capacity (%d). Consider lowering min_capacity!" % (
+                ideal_capacity, self.resource['min_capacity']))
+        return current_capacity, new_capacity
+
+    def set_capacity(self, capacity):
+        if self.dry_run:
+            return
+        asg_client = boto3.client('autoscaling', region_name=self.resource['region'])
+        try:
+            ret = asg_client.update_auto_scaling_group(AutoScalingGroupName=self.resource['id'],
+                                                       DesiredCapacity=capacity)
+        except ClientError as e:
+            log.error("Error modifying ASG: {0}".format(e))
+            if e.response['Error']['Code'] == 'ValidationError':
+                log.warning("A validation error on an ASG update is usually caused by trying to update capacity "
+                            "outside min/max, which is ok if we're scaling down!")
+                return
+            raise FailSetResourceCapacity
+        return ret
 
+    def is_resource_cancelled(self):
+        return not self.asg
 
-def get_instances_from_ip(ip, instance_descriptions):
-    """Filter AWS instance_descriptions based on PrivateIpAddress
+    def get_instance_type_weights(self):
+        return StaticWeight()
 
-    :param ip: private IP of AWS instance.
-    :param instance_descriptions: list of AWS instance description dicts.
-    :returns: list of instance description dicts"""
-    instances = [instance for instance in instance_descriptions if instance['PrivateIpAddress'] == ip]
-    return instances
 
+class StaticWeight:
```

Comment:
I still don't really like this, but add some docs to explain what it's for and when it'll go away.